source/opt/decoration_manager.cpp

```diff
@@ -231,31 +231,78 @@ std::vector<T> DecorationManager::InternalGetDecorationsFor(
 
 void DecorationManager::ForEachDecoration(uint32_t id,
                                           uint32_t decoration,
-                                          std::function<void(const ir::Instruction&)> f) const {
-  auto decoration_list = id_to_decoration_insts_.find(id);
-  if (decoration_list != id_to_decoration_insts_.end()) {
-    for (const ir::Instruction* inst : decoration_list->second) {
-      switch (inst->opcode()) {
-        case SpvOpDecorate:
-          if (inst->GetSingleWordInOperand(1) == decoration) {
-            f(*inst);
-          }
-          break;
-        case SpvOpMemberDecorate:
-          if (inst->GetSingleWordInOperand(2) == decoration) {
-            f(*inst);
-          }
-          break;
-        case SpvOpDecorateId:
-          if (inst->GetSingleWordInOperand(1) == decoration) {
-            f(*inst);
+                                          std::function<void(const ir::Instruction&)> f) {
+  for (const ir::Instruction* inst : GetDecorationsFor(id, false)) {
+    switch (inst->opcode()) {
+      case SpvOpDecorate:
+        if (inst->GetSingleWordInOperand(1) == decoration) {
+          f(*inst);
+        }
+        break;
+      case SpvOpMemberDecorate:
+        if (inst->GetSingleWordInOperand(2) == decoration) {
+          f(*inst);
+        }
+        break;
+      case SpvOpDecorateId:
+        if (inst->GetSingleWordInOperand(1) == decoration) {
+          f(*inst);
+        }
+        break;
+      default:
+        assert(false && "Unexpected decoration instruction");
+    }
+  }
+}
+
+void DecorationManager::CloneDecorations(uint32_t from, uint32_t to) {
+  auto const decoration_list = id_to_decoration_insts_.find(from);
+  if (decoration_list == id_to_decoration_insts_.end()) return;
+  for (ir::Instruction* inst : decoration_list->second) {
```

Comment:
I just saw that it contains a potential bug. id_to_decoration_insts_ also contains group decorations which are not handled but raise an assertion.

Reply:
In this case, you want to rewrite the OpGroupDecorate and OpGroupMemberDecorate, which GetDecorations for will remove.  Makes sense.

Reply:
CloneDecorations() and ForEachDecoration() are completely independent. I changed ForEachDecoration, because it had unhandled cases. I think, the intention was to have a function which goes through all (direct and indirect) specified Op(Member)Decorations and not group decorations. So, I fixed it.

Reply:
I see. Shall I put it in a second PR?